nomad/drainer/drain_testing.go

```diff
@@ -1,52 +1,142 @@
+// this channel will block anywhere else.)
+func (m *MockJobWatcher) Migrated() <-chan []*structs.Allocation {
+	return m.migratedCh
+}
+
+type MockDeadlineNotifier struct {
+	expiredCh <-chan []string
+	nodes     map[string]struct{}
+	sync.Mutex
 }
 
-func (m *MockNodeTracker) Update(node *structs.Node) {
+// NextBatch returns the channel of expired nodes. Tests can send on this
+// channel to simulate timer events in the NodeDrainer watch loop. (Sending on
+// this channel will block anywhere else.)
+func (m *MockDeadlineNotifier) NextBatch() <-chan []string {
+	return m.expiredCh
+}
+
+// Remove removes the given node from being tracked for a deadline.
+func (m *MockDeadlineNotifier) Remove(nodeID string) {
 	m.Lock()
 	defer m.Unlock()
-	m.Nodes[node.ID] = node
-	m.Events = append(m.Events, &MockNodeTrackerEvent{NodeUpdate: node})
+	delete(m.nodes, nodeID)
 }
 
-func (m *MockNodeTracker) events() []*MockNodeTrackerEvent {
+// Watch marks the node as being watched; this mock throws out the timer in lieu
+// of manully sending on the channel to avoid racy tests.
+func (m *MockDeadlineNotifier) Watch(nodeID string, _ time.Time) {
 	m.Lock()
 	defer m.Unlock()
+	m.nodes[nodeID] = struct{}{}
+}
+
+type MockRaftApplierShim struct {
+	state *state.StateStore
+}
+
+// AllocUpdateDesiredTransition mocks a write to raft as a state store update
+func (m *MockRaftApplierShim) AllocUpdateDesiredTransition(
+	allocs map[string]*structs.DesiredTransition, evals []*structs.Evaluation) (uint64, error) {
+	index, _ := m.state.LatestIndex()
+	index++
+	err := m.state.UpdateAllocsDesiredTransitions(structs.MsgTypeTestSetup, index, allocs, evals)
+	return index, err
+}
+
+// NodesDrainComplete mocks a write to raft as a state store update
+func (m *MockRaftApplierShim) NodesDrainComplete(
+	nodes []string, event *structs.NodeEvent) (uint64, error) {
+	index, _ := m.state.LatestIndex()
+	index++
+
+	updates := make(map[string]*structs.DrainUpdate, len(nodes))
+	nodeEvents := make(map[string]*structs.NodeEvent, len(nodes))
+	update := &structs.DrainUpdate{}
+	for _, node := range nodes {
+		updates[node] = update
+		if event != nil {
+			nodeEvents[node] = event
+		}
+	}
+	now := time.Now().Unix()
+
+	err := m.state.BatchUpdateNodeDrain(structs.MsgTypeTestSetup, index, now,
+		updates, nodeEvents)
+
+	return index, err
+}
+
+func testNodeDrainWatcher(t *testing.T) (*nodeDrainWatcher, *state.StateStore, *NodeDrainer) {
+	t.Helper()
+	store := state.TestStateStore(t)
+	limiter := rate.NewLimiter(100.0, 100)
+	logger := testlog.HCLogger(t)
+
+	drainer := &NodeDrainer{
+		enabled: false,
+		logger:  logger,
+		nodes:   map[string]*drainingNode{},
+		jobWatcher: &MockJobWatcher{
+			drainCh:    make(chan *DrainRequest),
+			migratedCh: make(chan []*structs.Allocation),
+			jobs:       map[structs.NamespacedID]struct{}{},
+		},
+		deadlineNotifier: &MockDeadlineNotifier{
+			expiredCh: make(<-chan []string),
```

Comment:
Hm, not sure I understand the point of drainCh, migratedCh, and expiredCh. They don't seem to ever have any writes, so they're all effectively nil chans?